src/Analyzers/CSharp/CodeFixes/UseCollectionExpression/CSharpCollectionExpressionRewriter.cs

```diff
@@ -0,0 +1,592 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CSharp.Analyzers.UseCollectionExpression;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.CSharp.UseCollectionInitializer;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Indentation;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.UseCollectionExpression;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.UseCollectionExpression;
+
+using static CSharpCollectionInitializerRewriter;
+using static SyntaxFactory;
+
+internal static class CSharpCollectionExpressionRewriter
+{
+    /// <summary>
+    /// Creates the final collection-expression <c>[...]</c> that will replace the given <paramref
+    /// name="expressionToReplace"/> expression.
+    /// </summary>
+    public static async Task<CollectionExpressionSyntax> CreateCollectionExpressionAsync<TParentExpression>(
+        Document workspaceDocument,
+        CodeActionOptionsProvider fallbackOptions,
+        TParentExpression expressionToReplace,
+        ImmutableArray<CollectionExpressionMatch> matches,
+        Func<TParentExpression, InitializerExpressionSyntax?> getInitializer,
+        Func<TParentExpression, InitializerExpressionSyntax, TParentExpression> withInitializer,
+        CancellationToken cancellationToken)
+        where TParentExpression : ExpressionSyntax
```

Comment:
this is the same as the original code, just updated to not have to take an ObjectCreationExpression as the parent.